README.md
# Pipe
Swift pipe operator similar to Elixir's and F#'s `|>` operator.


CHANGELOG.md

# 0.2.0
1. [[PIE-1](https://github.com/danthorpe/Pipe/pull/1)]: Adds more curried functions for Array and Sequence manipulation. This includes `sum` (for numeric types), `sort`, `find`, `insert` and `remove`.
1. [[PIE-2](https://github.com/danthorpe/Pipe/pull/2)]: Whoops - removes a mistaken `Float80` implementation of `sum`.


# 0.1.0
Initial Release - bare minimal implementation of `|>` and `map`, `filter` and `reduce`.


test.sh
#!/usr/bin/env bash
source /usr/local/opt/chruby/share/chruby/chruby.sh
chruby ruby
echo ''gem: --no-ri --no-rdoc'' > ~/.gemrc
bundle install
pushd "Example"
pod install
xcodebuild -workspace Pipe\ Dream.xcworkspace -scheme Pipe\ Dream -sdk iphonesimulator build test




Source/Pipe.swift
//
//  Pipe.swift
//
//  Created by Daniel Thorpe on 11/05/2015.
//

import Foundation

infix operator |> { precedence 50 associativity left }

/**
Pipe. e.g. x |> f is transformed into f(x). But it's chainable
so that x |> f |> g |> h is transformed into h(g(f(x)))
*/
public func |> <Input, Output>(lhs: Input, rhs: Input -> Output) -> Output {
    return rhs(lhs)
}

/// Curried version of map
public func map<S: SequenceType, T>(transform: (S.Generator.Element) -> T)(source: S) -> [T] {
    return source.map(transform)
}

/// Curried version of filter
public func filter<S: SequenceType>(includeElement: (S.Generator.Element) -> Bool)(source: S) -> [S.Generator.Element] {
    return source.filter(includeElement)
}

/// Curried version of reduce
public func reduce<S: SequenceType, U>(initial: U, combine: (U, S.Generator.Element) -> U)(source: S) -> U {
    return source.reduce(initial, combine: combine)
}

/// Curried version of sorted
public func sorted<S: SequenceType>(isOrderedBefore: (S.Generator.Element, S.Generator.Element) -> Bool)(source: S) -> [S.Generator.Element] {
    return source.sort(isOrderedBefore)
}

/// Curried version of find
public func find<C: CollectionType where C.Generator.Element: Equatable>(value: C.Generator.Element) -> (C) -> C.Index? {
    return { $0.indexOf(value) }
}

/// Curried version Array.append
public func append<T>(element: T) -> [T] -> [T] {
    return { (var array) in
        array.append(element)
        return array
    }
}

/// Curried version Array.insert:atIndex:
public func insert<T>(element: T, atIndex index: Int) -> [T] -> [T] {
    return { (var array) in
        array.insert(element, atIndex: index)
        return array
    }
}

/// Curried version Array.removeAtIndex:
public func removeAtIndex<T>(index: Int) -> [T] -> [T] {
    return { (var array) in
        array.removeAtIndex(index)
        return array
    }
}

/// Curried version of sum
public func sum<S: SequenceType where S.Generator.Element == UInt8>(source: S)   -> S.Generator.Element { return source.reduce(0, combine: +) }
public func sum<S: SequenceType where S.Generator.Element == Int8>(source: S)    -> S.Generator.Element { return source.reduce(0, combine: +) }
public func sum<S: SequenceType where S.Generator.Element == UInt16>(source: S)  -> S.Generator.Element { return source.reduce(0, combine: +) }
public func sum<S: SequenceType where S.Generator.Element == Int16>(source: S)   -> S.Generator.Element { return source.reduce(0, combine: +) }
public func sum<S: SequenceType where S.Generator.Element == UInt32>(source: S)  -> S.Generator.Element { return source.reduce(0, combine: +) }
public func sum<S: SequenceType where S.Generator.Element == Int32>(source: S)   -> S.Generator.Element { return source.reduce(0, combine: +) }
public func sum<S: SequenceType where S.Generator.Element == UInt64>(source: S)  -> S.Generator.Element { return source.reduce(0, combine: +) }
public func sum<S: SequenceType where S.Generator.Element == Int64>(source: S)   -> S.Generator.Element { return source.reduce(0, combine: +) }
public func sum<S: SequenceType where S.Generator.Element == UInt>(source: S)    -> S.Generator.Element { return source.reduce(0, combine: +) }
public func sum<S: SequenceType where S.Generator.Element == Int>(source: S)     -> S.Generator.Element { return source.reduce(0, combine: +) }
public func sum<S: SequenceType where S.Generator.Element == Double>(source: S)  -> S.Generator.Element { return source.reduce(0, combine: +) }
public func sum<S: SequenceType where S.Generator.Element == Float>(source: S)   -> S.Generator.Element { return source.reduce(0, combine: +) }
public func sum<S: SequenceType where S.Generator.Element == CGFloat>(source: S) -> S.Generator.Element { return source.reduce(0, combine: +) }


Example/Pipe Dream/AppDelegate.swift
//
//  AppDelegate.swift
//  Pipe Dream
//
//  Created by Daniel Thorpe on 15/05/2015.
//  Copyright (c) 2015 Dan Thorpe. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



Example/Pipe DreamTests/PipeTests.swift
//
//  PipeTests.swift
//  Pipe Dream
//
//  Created by Daniel Thorpe on 08/06/2015.
//  Copyright © 2015 Dan Thorpe. All rights reserved.
//

import UIKit
import XCTest
import Pipe


class PipeTests: XCTestCase {

    let numbers = [4, 8, 15, 16, 23, 42]

    func test_Pipe_map() {
        XCTAssertEqual(numbers |> map { $0 * 10 }, [40, 80, 150, 160, 230, 420])
    }

    func test_Pipe_filter_then_map() {
        let result = numbers
            |> filter { $0 % 2 != 0 }
            |> map { $0 * 10 }
        XCTAssertEqual(result, [150, 230])
    }

    func test_Pipe_map_then_filter() {
        let result = numbers
            |> map { $0 * 7 } // [28, 56, 105, 112, 161, 294]
            |> filter { $0 % 2 != 0 }
        XCTAssertEqual(result, [105, 161])
    }

    func test_Pipe_reduce() {
        XCTAssertEqual(numbers |> reduce(0, combine: +), 108)
    }

    func test_Pipe_sum_integers() {
        XCTAssertEqual(numbers |> sum, 108)
    }

    func test_Pipe_sum_doubles() {
        XCTAssertEqual(numbers |> map { Double($0) } |> sum, 108.0)
    }

    func test_Pipe_sum_floats() {
        XCTAssertEqual(numbers |> map { Float($0) } |> sum, 108.0)
    }

    func test_Pipe_sort_reverse() {
        XCTAssertEqual(numbers |> sorted(>), [42, 23, 16, 15, 8, 4])
    }

    func test_Pipe_sort_again() {
        XCTAssertEqual(numbers |> sorted(>) |> sorted(<), numbers)
    }

    func test_Pipe_custom_closure() {
        XCTAssertEqual(numbers |> { s in s.map { $0 * 10 } }, [40, 80, 150, 160, 230, 420])
    }

    func test_Pipe_find_a_number() {
        let result = numbers |> find(15)
        XCTAssertTrue(result == 2)
    }

    func test_Pipe_find_return_none() {
        let result = numbers |> find(666)
        XCTAssertTrue(result == nil)
    }

    func test_Pipe_array_remove_insert() {
        let result = numbers
            |> removeAtIndex(0)
            |> insert(6, atIndex: 0)
        XCTAssertEqual(result[0], 6)
    }
}




Example/Pipe DreamTests/PipeSpecs.swift
//
//  PipeSpec.swift
//
//  Created by Daniel Thorpe on 11/05/2015.
//

import UIKit
import XCTest
import Quick
import Nimble
import Pipe

class PipeSpecs: QuickSpec {
    override func spec() {
        describe("Pipe") {
            let numbers = [4, 8, 15, 16, 23, 42]

            describe("map") {
                it("should be 10x") {
                    let result = numbers
                        |> map { $0 * 10 }
                    expect(result).to(equal([40, 80, 150, 160, 230, 420]))
                }
            }

            describe("filter then map") {
                it("should have only two odd numbers") {
                    let result = numbers
                        |> filter { $0 % 2 != 0 }
                        |> map { $0 * 10 }
                    expect(result).to(equal([150, 230]))
                }
            }

            describe("map then filter") {
                it("should have only two odd numbers") {
                    let result = numbers
                        |> map { $0 * 7 } // [28, 56, 105, 112, 161, 294]
                        |> filter { $0 % 2 != 0 }

                    expect(result).to(equal([105, 161]))
                }
            }

            describe("reduce") {
                it("should sum") {
                    let result = numbers
                        |> reduce(0, +)
                    expect(result).to(equal(108))
                }
            }

            describe("sum") {
                it("should work with integers") {
                    let result = numbers |> sum
                    expect(result).to(equal(108))
                }

                it("should work with doubles") {
                    let result = numbers |> map { Double($0) } |> sum
                    expect(result).to(equal(108.0))
                }

                it("should work with floats") {
                    let result = numbers |> map { Float($0) } |> sum
                    expect(result).to(equal(108.0))
                }
            }

            describe("sort") {
                it("should reverse") {
                    let result = numbers
                        |> sorted(>)
                    expect(result).to(equal([42, 23, 16, 15, 8, 4]))
                }

                it("sorting again") {
                    let result = numbers
                        |> sorted(>)
                        |> sorted(<)
                    expect(result).to(equal(numbers))
                }
            }

            describe("custom closure") {
                it("should be 10x") {
                    let result = numbers
                        |> { s in map(s) { $0 * 10 } }
                    expect(result).to(equal([40, 80, 150, 160, 230, 420]))
                }
            }

            describe("find") {
                it("should find a number") {
                    let result = numbers
                        |> find(15)
                    expect(result).to(equal(2))
                }

                it("should return .None if not found") {
                    let result = numbers
                        |> find(666)
                    expect(result).to(beNil())
                }
            }

            describe("array functions") {
                it("remove and insert") {
                    let result = numbers
                        |> removeAtIndex(0)
                        |> insert(6, atIndex: 0)
                    expect(result[0]).to(equal(6))
                }
            }
        }
    }
}
